### PR TITLE
[Test] Shard "Small & Large" tests

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
The `Small & Large` build on buildkite runs more than 2hr until they finish. This PR shard the build into smaller units so they can run in parallel and finish faster.

Also, use positive tag filtering instead of negative tag filtering for small and large tests. With negative tag filtering, `Small & Large` unintentionally include tests that are part of other builds, e.g. debug build tests, client tests and asan tests.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simon-mo — CHANGES_REQUESTED** [review]
Thanks for doing this. Please hold on merging as we need more capacity for macOS runners.

**Comment by simon-mo** [comment]
And for sharding, I would encourage you to try out the parallelism flag like this
https://github.com/ray-project/ray/blob/7efe1bef114053c3b5168160ed88af2f40dcebf8/.buildkite/pipeline.windows.yml#L29
and
https://github.com/ray-project/ray/blob/master/ci/travis/ci.sh#L180-L194
which uses 
https://github.com/ray-project/ray/blob/master/ci/travis/bazel-sharding.py
under the hood.

This should make the job spec much cleaner. 

If you can just do this for linux that would be awesome and we can merge almost as-is (but using the native parallelism option will be appreciated as it clean up the code). We can fix the macOS sharding once the capacity is ramped up.

**Comment by mwtian** [comment]
I moved sharding to `ci.sh`, although I'm keeping the manual sharding for large tests. @simon-mo can you help deploy [this](https://github.com/ray-project/buildkite-ci-stack/commit/44ddf0f037b5ebabaf2c5ce6b6037e5e4e66dab7) which seems needed to propagate the `BUILDKITE_PARALLEL_JOB` to the inner docker?

**Comment by simon-mo** [comment]
Done.

File: .buildkite/pipeline.macos.yml [hunk 1]
```diff
@@ -79,15 +79,23 @@ steps:
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/travis/upload_build_info.sh; fi }; trap cleanup EXIT
     - ./ci/travis/ci.sh test_cpp
 
-- label: ":mac: :apple: Small and Large"
+- label: ":mac: :apple: Small & Client"
```
Comment (simon-mo) [inline .buildkite/pipeline.macos.yml:82]: Can you revert the change for macos? We can add them back when we have more capacity.

File: ci/travis/ci.sh [hunk 2]
```diff
@@ -193,6 +194,23 @@ test_python() {
   fi
 }
 
+# For running large Python tests on Linux and MacOS.
+test_large() {
+  bazel test --config=ci "$(./scripts/bazel_export_options)" --test_env=CONDA_EXE --test_env=CONDA_PYTHON_EXE \
```
Comment (simon-mo) [inline ci/travis/ci.sh:199]: You still need to run bazel-sharding.py to list and split the targets in here. 

Also you don't have to put them here. Just write them in yaml works too.

**Review by simon-mo — APPROVED** [review]

File: .buildkite/pipeline.macos.yml (hunk 1 above)
Comment (mwtian) [reply .buildkite/pipeline.macos.yml:82]: Or how about we wait until there is more capacity? Otherwise with a slow MacOS build step the overall build is still going to be slow..

File: ci/travis/ci.sh (hunk 2 above)
Comment (mwtian) [reply ci/travis/ci.sh:199]: This uses per-shard tag filter, why is `bazel-sharding.py` still needed? Putting the command here reduces duplication a little bit, but I'm open to move the bazel command back to yaml.

Comment (mwtian) [reply ci/travis/ci.sh:199]: I agree in future, probably we can convert all Linux Python tests to use the `bazel-sharding.py` script. But it will be a larger change.

**Comment by mwtian** [comment]
Looks like Mac builds have a shorter queue now. If there is no objection, we can try to merge this tomorrow.

File: ci/travis/ci.sh (hunk 2 above)
Comment (simon-mo) [reply ci/travis/ci.sh:199]: Ah i missed the `--test_tag_filters="large_size_python_tests_shard_${BUILDKITE_PARALLEL_JOB}"`. Thanks for explaining!

**Comment by iycheng** [comment]
I really like this change. It make small/large test run faster.
